improc/zogy.py

```diff
@@ -336,35 +337,66 @@ def zogy_add_weights_flags(ref_weight, new_weight, ref_flags, new_flags, ref_psf
     ----------
     ref_weight: numpy.ndarray
         The weight image of the reference image.
+
     new_weight: numpy.ndarray
         The weight image of the new image.
+
     ref_flags: numpy.ndarray
         The flag image of the reference image.
+
     new_flags: numpy.ndarray
         The flag image of the new image.
+
+    ref_zp: float
+         Zeropoint for the reference (m=-2.5*log(data) + zp)
+
+    new_zp: float
+         Zeropoint for the new.
+
+    sub_zp: float
+         Zeropoint for the difference image (and the returned weights)
+
     ref_psf_fwhm: float
         The FWHM of the PSF of the reference image, in pixels!
+
     new_psf_fwhm: float
         The FWHM of the PSF of the new image, in pixels!
 
     Returns
     -------
     outwt: numpy.ndarray
-        The combined weight image.
+        The combined weight image, with effective magnitude zeropoint sub_zp.
+
     outfl: numpy.ndarray
         The combined flag image.
     """
-    # combine the weights
-    mask = (ref_weight == 0) | (new_weight == 0)
+
+    # Turn weights into variances
     # divide without dividing by zero (ref: https://stackoverflow.com/a/37977222)
     w1 = np.divide(1, ref_weight, out=np.zeros_like(ref_weight), where=ref_weight != 0)
     w2 = np.divide(1, new_weight, out=np.zeros_like(new_weight), where=new_weight != 0)
-    outwt = np.divide(1, w1 + w2, out=np.zeros_like(w1), where=(mask != 0) & ((w1 + w2) != 0))
+
+    # Scale the variances so that they're on the same zeropoint as the sub image.
+    # variances have units equivalent to i², where i is a data pixel value in adu.
+    #
+    # For two fluxes to have the same magnitude:
+    #   m = -2.5*log(f1) + zp1 = -2.5*log(f2) + zp2
+    #   -2.5*log(f1/f2) = zp2 - zp1
+    #   f1/f2 = 10**(-0.4*(zp2-zp1))
+    #   σ1²/σ2² = 10**(-0.8*(zp2-zp1))
+    w1 *= 10 ** ( -0.8 * ( ref_zp - sub_zp ) )
+    w2 *= 10 ** ( -0.8 * ( new_zp - sub_zp ) )
+
+    # combine the weights
+    mask = (ref_weight == 0) | (new_weight == 0) | ( new_flags !=0 ) | ( ref_flags != 0 )
```

Comment:
If I understand right, the `|` statements are applying bitwise OR, applied individually across each element of the array? Like the output of `mask` is an array of same size with `1` (or `True`?) for any element where that location was equal to 0 in one of these 4 arrays?

Thats a pretty neat and compact way of doing this with numpy arrays.